File: Desktop/DSAN_5000/HW-02/codes-and-outputs/analytical.R
# Import libraries
library(tidyverse)
library(reshape2)

# Read the CSV files
loans = read.csv("loans_r.csv")
accounts = read.csv("../data/accounts.csv")
cards = read.csv("../data/cards.csv")
districts = read.csv("../data/districts.csv")
links = read.csv("../data/links.csv")
transactions = read.csv("../data/transactions.csv")

# Rename columns in dataframes
names(accounts)[names(accounts) == "id"] = "account_id"
names(accounts)[names(accounts) == "date"] = "open_date"
names(districts)[names(districts) == "id"] = "district_id"

# Merge the 'districts' dataframe with the 'accounts' dataframe
accounts = merge(accounts, districts[, c("district_id", "name")], by.x = "district_id", by.y = "district_id", all = TRUE)

# Rename the merged column to 'district_name'
names(accounts)[names(accounts) == "name"] = "district_name"

# Drop the 'district_id' column
accounts = accounts[, !(names(accounts) %in% c("district_id"))]

# Drop columns "id" and "date" from the 'loans' dataframe
loans = loans[, !(names(loans) %in% c("id", "date"))]

# Create empty vectors to store "status" and "default" values
status = vector("character", length = nrow(loans))
default = vector("logical", length = nrow(loans))

# Iterate through the "status" column and populate "status" and "default" vectors
for (i in 1:nrow(loans)) {
  if (loans$status[i] == "A" ){
    status[i] = "expired"
    default[i] = FALSE
  }  else if (loans$status[i] == "B"){
    status[i] = "expired"
    default[i] = TRUE    
  } else if (loans$status[i] == "C" || loans$status[i] == "D") {
    status[i] = "current"
    default[i] = FALSE
  }
}

# Replace the "status" and "default" columns in the 'loans' dataframe
loans$status = status
loans$loan_default = default

# Rename columns
colnames(loans) = c("account_id","loan_amount", "loan_payments", "loan_term", "loan_status", "loan_default")

# Merge loans and accounts dataframe
accounts = merge(accounts, loans, all = TRUE)

# Create a new column 'loan' in the 'accounts' dataframe
accounts$loan = NA
for (i in 1:nrow(accounts)) {
  if (is.na(accounts$loan_amount[i])) {
    accounts$loan[i] = FALSE
  } else {
    accounts$loan[i] = TRUE
  }
}

# Drop the "id" column from the 'links' dataframe
links = links[, !(names(links) %in% c("id"))]

# Drop the "id" and "type" columns from the 'cards' dataframe
cards = cards[, !(names(cards) %in% c("id", "type"))]

# Rename the "link_id" column to "client_id" in the 'cards' dataframe
colnames(cards)[colnames(cards) == "link_id"] = "client_id"

# Merge the 'links' and 'cards' dataframes 
a = merge(links, cards, all = TRUE)

# Create a cross-tabulation to count the number of customers per account_id
b = table(a$account_id)

# Convert the result to a dataframe and reset the index
b = as.data.frame(b)
names(b) = c("account_id", "num_customers")

# Add 'num_customers' column by merging the 'accounts' dataframe with the 'b' dataframe
accounts = merge(accounts, b, by = "account_id", all = TRUE)

# Create a subset of 'a' dataframe and count the number of credit cards per account_id
c = table(a[!is.na(a$issue_date), "account_id"])

# Convert the result to a dataframe and reset the index
c = as.data.frame(c)
names(c) = c("account_id", "credit_cards")

# Add 'credit_cards' column by merging the 'accounts' dataframe with the 'c' dataframe
accounts = merge(accounts, c, by = "account_id", all = TRUE)

# Replace missing values in the 'credit_cards' column with 0
accounts$credit_cards[is.na(accounts$credit_cards)] = 0

# Initialize columns
accounts$max_balance = NA
accounts$min_balance = NA
accounts$max_withdrawal = NA
accounts$min_withdrawal = NA

# Create a subset of 'transactions' dataframe for debit transactions
withdraw = transactions %>% filter(type == "debit")

# Iterate through unique account IDs in 'transactions'
for (i in unique(transactions$account_id)) {
  # Calculate the max and min balances for the current account ID
  min_balance = min(transactions$balance[transactions$account_id == i])
  max_balance = max(transactions$balance[transactions$account_id == i])
  
  # Update the 'accounts' dataframe with max and min balances
  accounts$max_balance[accounts$account_id == i] = max_balance
  accounts$min_balance[accounts$account_id == i] = min_balance

  # Calculate the max and min withdrawal amounts for the current account ID
  min_withdrawal = min(withdraw$amount[withdraw$account_id == i])
  max_withdrawal = max(withdraw$amount[withdraw$account_id == i])
  
  # Update the 'accounts' dataframe with max and min withdrawal amounts
  accounts$max_withdrawal[accounts$account_id == i] = max_withdrawal
  accounts$min_withdrawal[accounts$account_id == i] = min_withdrawal
}

# Create a subset of 'withdraw' dataframe for transactions with "credit card" method
payments = withdraw[withdraw$method == "credit card", ]

# Initialize a column for credit card payments in the 'accounts' dataframe
accounts$cc_payments = 0

# Iterate through unique account IDs in 'transactions'
for (i in unique(transactions$account_id)) {
  # Count the number of credit card payments for the current account ID
  num_cc_payments = sum(payments$account_id == i)
  
  # Update the 'accounts' dataframe with the number of credit card payments
  accounts$cc_payments[accounts$account_id == i] = num_cc_payments
}

# Reorder the columns
accounts = accounts %>%
  select('account_id', 'district_name', 'open_date', 'statement_frequency','num_customers','credit_cards', 'loan', 'loan_amount','loan_payments', 'loan_term', 'loan_status', 'loan_default','max_withdrawal', 'min_withdrawal', 'cc_payments', 'max_balance', 'min_balance')

# Save the resulting dataframe to a CSV file
write.csv(accounts, file = "./analytical_r.csv", row.names = FALSE)

File: Desktop/DSAN_5000/HW-02/codes-and-outputs/analytical.py
# Import libraries
import pandas as pd
import numpy as np

# Read the CSV files
loans = pd.read_csv("loans_py.csv")
accounts = pd.read_csv("../data/accounts.csv")
cards = pd.read_csv("../data/cards.csv")
districts = pd.read_csv("../data/districts.csv")
links = pd.read_csv("../data/links.csv")
transactions = pd.read_csv("../data/transactions.csv")

# Rename columns in dataframes
accounts = accounts.rename(columns = {"id":"account_id"})
accounts = accounts.rename(columns = {"date":"open_date"})
districts = districts.rename(columns = {"id":"district_id"})

# Add "district_name" column
accounts = accounts.merge(districts.loc[:,["district_id","name"]], how='outer')
accounts = accounts.rename(columns = {"name":"district_name"})
accounts = accounts.drop(columns="district_id", axis=1)

# Add "loan_amount","loan_payments","loan_term","loan_status","loan_default" column
loans = loans.drop(columns=["id","date"], axis=1)
default = []
status = []
for i in loans["status"]:
    if i == "A":
        status.append("expired")
        default.append(False)
    if i == "B":
        status.append("expired")
        default.append(True)
    if i == "C":
        status.append("current")
        default.append(False)
    if i == "D":
        status.append("current")
        default.append(False)
loans["status"] = status
loans["defualt"] = default
loans = loans.rename(columns={"amount":"loan_amount","payments":"loan_payments","term":"loan_term","status":"loan_status","defualt":"loan_default"})
accounts = accounts.merge(loans, how='outer')

# Add "loan" column
for i in accounts.index:
    if accounts.loc[i,"loan_amount"].is_integer():
        accounts.loc[i,"loan"] = True
    else:
        accounts.loc[i,"loan"] = False

# Initialize dataframes for merging
links = links.drop(columns = "id")
cards = cards.drop(columns = ["id","type"])
cards = cards.rename(columns={"link_id":"client_id"})

# Add "num_customer" column
clients = links.merge(cards, how="outer")
num_customers = pd.crosstab(index = clients["account_id"], columns="num_customers").reset_index()
accounts = accounts.merge(num_customers, how='outer')

# Add "credit_cards" column
credit_cards = pd.crosstab(index = clients.dropna(subset=['issue_date'])["account_id"], columns="credit_cards").reset_index()
accounts = accounts.merge(credit_cards, how='outer')
accounts['credit_cards'] = accounts['credit_cards'].fillna(0)

withdraw = transactions[transactions["type"]=="debit"]
payments = withdraw[withdraw["method"]=="credit card"]
# Loop through each account_id
for i in transactions["account_id"].unique():
    # Add "min_balance" and "max_balance" columns
    min_balance = transactions[transactions["account_id"]==i]["balance"].min()
    max_balance = transactions[transactions["account_id"]==i]["balance"].max()
    accounts.loc[accounts["account_id"]==i,"max_balance"] = max_balance
    accounts.loc[accounts["account_id"]==i,"min_balance"] = min_balance
    
    # Add "min_withdrawal" and "max_withdrawal" columns
    min_amount = withdraw[withdraw["account_id"]==i]["amount"].min()
    max_amount = withdraw[withdraw["account_id"]==i]["amount"].max()
    accounts.loc[accounts["account_id"]==i,"max_withdrawal"] = max_amount
    accounts.loc[accounts["account_id"]==i,"min_withdrawal"] = min_amount

    # Add "cc_payments" column
    accounts.loc[accounts["account_id"]==i,"cc_payments"] = len(payments[payments["account_id"]==i])

# Make dataframe tidy
accounts = accounts[['account_id', 'district_name', 'open_date', 'statement_frequency','num_customers','credit_cards', 'loan', 'loan_amount','loan_payments', 'loan_term', 'loan_status', 'loan_default','max_withdrawal', 'min_withdrawal', 'cc_payments', 'max_balance', 'min_balance']]
accounts = accounts.sort_values(by='account_id', ascending=True)

# Save the resulting DataFrame to a CSV file
accounts.to_csv("./analytical_py.csv", index=False)